### ISSUE TITLE
Aggregation queries

### ISSUE THREAD

**Issue by antessio** [antessio/dynamoplus#35]
In order to support aggregated queries (i.e.: count, max, min, ...), define a new kind of index (and query as well). 

The aggregation query should be flexible. 

So the index declaration should allow the user to provide all the informations to aggregate the results.
For example:
- given a collection `restaurant` and `reviews`, i want to get for a given `restaurant`  the review rates average. 
- given a collection `restaurant` and `reviews`, i want to get for a given `restaurant`  the reviews count
- given a collection `restaurant` and `reviews`, i want to get for a given `restaurant`  the max review rate

For aggregation on the whole collection the value should be stored as a collection attribute. 
For example: 
- given a collection `products`, I want to get the count of all the products
- given a collection `products`, I want to get the average product price
- given a collection `products`, I want to get the minimum product expiration

Basically these are two different aggregation models. 

To perform the aggregations, when the dynamo db streams handler intercepts a change and an aggregation index has been defined, it should update the aggregation value accordingly (i.e.: calculate the maximum, calculate the minimum, increase the count). 
In the first example the aggregation should be triggered on the changes performed on `reviews` collection and should have `restaurant` as target. 
In the second, the aggregation should be triggered on `restaurant` collection and should have the same collection as target.

**Comment by antessio** [comment]
# Aggregated index draft

All the aggregates indexes have complexity O(1), they usually take advantage of redundant data stored on either the collection or the index itself. 


## Create

`POST /index/`

### Collection count

```json

{
  "collection": {
    "name": "restaurant"
  },
  "type": "AGGREGATION",
  "aggregation": {
    "on": ["INSERT"],
    "type": "COLLECTION_COUNT"
  } 
}

```

On `INSERT` on get the value `count`  from `collection#restaurant` and increment it
then update the field `count` in `collection#restaurant`  


### Average aggregation

```json

{
  "collection": {
    "name": "review"
  },
  "type": "AGGREGATION",
  "aggregation": {
    "on": ["INSERT"],
    "type": "AVG",
    "target_field": "rate"
        
  } 
}

```


On the event `INSERT` on the collection `review`, get the values `rate_count` and `rate_sum` from `collection#review` 
then calculate the average value:
`review_avg_rate=(rate_sum+review.rate)/(review_count+1)` then update the values of `rate_count`, `rate_sum` and `rate_avg` accordingly


### Average aggregation with conditions

```json

{
  "collection": {
    "name": "restaurant"
  },
  "type": "AGGREGATION",
  "aggregation": {
    "on": ["INSERT"],
    "type": "AVG",
    "target_field": "seat",
	"matches":{
		"and":[
			{
				"eq": { 
					"field_name":"type",
					"value":"pizzeria"
				}
			}
		]
	}
  } 
}

```



On the event `INSERT` on the collection `restaurant`, if the document matches the query in `matches` then get the values `seat_count` and `seat_sum` from `type_pizzeria_avg_seat`
then calculate the average value:
`type_pizzeria_avg_seat=(seat_sum+restaurant.seath)/(seat_count+1)` then update the values of `seat_count`, `seat_sum` and `type_pizzeria_avg_seat` accordingly

### Average joined aggregation 

```json

{
  "collection": {
    "name": "review"
  },
  "type": "AGGREGATION",
  "aggregation": {
    "on": ["INSERT"],
    "join": {
        "collection_name": "restaurant",
        "using_field": "restaurant_id"
    },
    "type": "AVG_JOIN",
    "target_field": "rate"
        
  } 
}

```


On the event `INSERT` on the collection `review`, get the values `rate_count` and `rate_sum` from `restaurant#avg_review_rate` then calculate the average value: 
`avg_review_rate=(rate_sum+review.rate)/(rate_count+1)` then update the values of `rate_count`, `rate_sum` and `restaurant#avg_review_rate` accordingly


### Sum aggregation 

`POST /index/`


```json

{
  "collection": {
    "name": "booking"
  },
  "type": "AGGREGATION",
  "aggregation": {
    "on": ["INSERT"],
    "join": {
        "collection_name": "restaurant",
        "using_field": "restaurant_id"
    },
    "type": "SUM",
    "target_field": "amount"
        
  } 
}

```


On the event `INSERT` on the collection `booking`, find the row `restaurant#sum_booking_amount` and read its value
sum the value with `booking.amount` then update `restaurant#sum_booking_amount` with the new value.



### Sum count aggregation 

`POST /index/`


```json

{
  "collection": {
    "name": "booking"
  },
  "type": "AGGREGATION",
  "aggregation": {
    "on": ["INSERT"],
    "join": {
        "collection_name": "restaurant",
        "using_field": "restaurant_id"
    },
    "type": "SUM_COUNT"
        
  } 
}

```

Depends on `COLLECTION_COUNT`. 

On the event `INSERT` on the collection `booking`, find the row `restaurant#count_booking`, read its value
and update the value with the value `count` from `collection#booking` then update `restaurant#count_booking` with the new value.



### Min/max aggregation 

`POST /index/`


```json

{
  "collection": {
    "name": "booking"
  },
  "type": "AGGREGATION",
  "aggregation": {
    "on": ["INSERT"],
    "join": {
        "collection_name": "restaurant",
        "using_field": "restaurant_id"
    },
    "type": "MIN",
    "target_field": "amount"
        
  } 
}

```


On the event `INSERT` on the collection `booking`, find the row `restaurant#min_booking_amount` and read its value
then compare it with the current created `booking.amount` if it's lower update the value of `restaurant#min_booking_amount`
otherwise nothing. Similar with `MAX`



## Query

`GET /<collection-name>/query?aggregation=<aggregation_id>`

Examples:
- `GET /restaurant/query?aggregation_type=COLLECTION_COUNT`
- `GET /review/query?aggregation=AVG&id=avg_rate`
- `GET /restaurant/query?aggregation=AVG&id=type_pizzeria_avg_seat`
- `GET /restaurant/query?aggregation=AVG_JOIN&id=avg_review_rate`
- `GET /restaurant/query?aggregation=SUM&id=sum_booking_amount`
- `GET /restaurant/query?aggregation=SUM_COUNT&id=count_booking`
- `GET /restaurant/query?aggregation=MIN&id=min_booking_amount`

**Comment by antessio** [comment]
Note: perform e2e tests on concurrencies

If two client update a resource the count should be updated consistently 
This should be guaranteed by using dynamo stream

**Comment by antessio** [comment]
## Aggregations

`GET /<collection-name>/aggregations/<aggregation_id>`


**Response Body**
```json
{
  "type": "<aggregation type>",
  "result": {
    "count": "<integer>",
    "sum": "<double>",
    "avg": "<double>",
    "min": "<double>",
    "max": "<double>"
  }
}
```

Note:

 - `count`, `sum`, `avg`, `min` and `max` depends on the aggregation type

### Examples

#### COLLECTION_COUNT



**Response Body**

```json
{
  "type": "COLLECTION_COUNT",
  "result": {
    "count": 10
  }
}
```

#### AVG

**Response Body**
```json
{
  "type": "AVG",
  "result": {
    "avg": 23.30
  }
}
```

#### SUM

**Response Body**
```json
{
  "type": "SUM",
  "result": {
    "sum": 2300
  }
}
```

#### MIN

**Response Body**
```json
{
  "type": "MIN",
  "result": {
    "min": 10
  }
}
```

#### MAX


**Response Body**
```json
{
  "type": "MAX",
  "result": {
    "min": 10200
  }
}
```
